bitmex-stream/src/lib.rs

```diff
@@ -1,31 +1,84 @@
+use std::time::UNIX_EPOCH;
 use tokio_tungstenite::tungstenite;
 use tracing::Instrument;
+use url::Url;
 
-pub use tokio_tungstenite::tungstenite::Error;
+pub use anyhow::Error;
 
 /// Connects to the BitMex websocket API, subscribes to the specified topics (comma-separated) and
 /// yields all messages.
 ///
-/// To keep the connection alive, a websocket `Ping` will be sent every 5 seconds in case no other
+/// To keep the connection alive, a websocket `Ping` is sent every 5 seconds in case no other
 /// message was received in-between. This is according to BitMex's API documentation: https://www.bitmex.com/app/wsAPI#Heartbeats
 pub fn subscribe<const N: usize>(
     topics: [String; N],
     network: Network,
 ) -> impl Stream<Item = Result<String, Error>> + Unpin {
+    subscribe_impl(topics, network, None)
+}
+
+/// Connects with authentication to the BitMex websocket API, subscribes to the
+/// specified topics (comma-separated) and yields all messages.
+///
+/// To keep the connection alive, a websocket `Ping` is sent every 5 seconds in case no other
+/// message was received in-between. This is according to BitMex's API documentation: https://www.bitmex.com/app/wsAPI#Heartbeats
+pub fn subscribe_with_credentials<const N: usize>(
+    topics: [String; N],
+    network: Network,
+    credentials: Credentials,
+) -> impl Stream<Item = Result<String, Error>> + Unpin {
+    subscribe_impl(topics, network, Some(credentials))
+}
+
+fn subscribe_impl<const N: usize>(
+    topics: [String; N],
+    network: Network,
+    credentials: Option<Credentials>,
+) -> impl Stream<Item = Result<String, Error>> + Unpin {
+    let url = network.to_url();
+    let url = format!("wss://{url}/realtime");
+
     let stream = stream! {
         tracing::debug!("Connecting to BitMex realtime API");
 
-        let subscription = topics.join(",");
-        let url = network.to_url();
-        let (mut connection, _) = tokio_tungstenite::connect_async(format!("wss://{url}/realtime?subscribe={subscription}")).await?;
+        let (mut connection, _) = tokio_tungstenite::connect_async(url.clone())
+            .await.context("Could not connect to websocket")?;
 
         tracing::info!("Connected to BitMex realtime API");
 
+        let mut messages = vec![];
+        if let Some(credentials) = credentials {
+            let start = SystemTime::now();
+            let expires = start
+                .duration_since(UNIX_EPOCH)?
+                .add(Duration::from_secs(5))
+                .as_secs();
+            let signature = credentials.sign(Method::GET, expires, &Url::parse(url.as_str())?, "");
+            messages.push(tungstenite::Message::try_from(signature.into_command())?);
+        }
+
+        messages.push(tungstenite::Message::try_from(Command::Subscribe(
+            topics.to_vec(),
+        ))?);
+
+        for message in messages {
+            let _ = connection
+                .send(message)
+                .await;
+        }
```

Comment:
🙃 Topics into an iterator of futures could be simpler.

Alternatively if you send the authentication message separately you could simply iterator through `topics` with a `for` loop here. That would get rid of `messages`.